### PR TITLE
Javascript application for fuzzing

### PR DESCRIPTION
**Pull request by turalsalamov**
It is a nodejs javascript demo application with vulnerability for fuzzing. Details of application are described in README.md file.

### CODE REVIEW

File: js_ts/nodejs-js/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Nodejs JS demo application
+
+## TODO application
+
+### Usage of application
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
## Functionality
```

File: js_ts/nodejs-js/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Nodejs JS demo application
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
# NodeJS demo application
```

File: js_ts/nodejs-js/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+## TODO application
+
+### Usage of application
+
+It is simple nodejs express TODO application which has several functionalities, such as, adding,
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
It is simple nodeJS express TODO application, which has several functionalities, such as, adding, 
```

File: js_ts/nodejs-js/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Nodejs JS demo application
+
+## TODO application
+
+### Usage of application
+
+It is simple nodejs express TODO application which has several functionalities, such as, adding, 
+deleting, listing TODOs, deleting whole json file and command execution in the server. The application creates
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
deleting single TODOs, and listing TODOs. Additional functionality includes, deleting the whole TODO list json file and command execution on the server. The application creates 
```

File: js_ts/nodejs-js/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Nodejs JS demo application
+
+## TODO application
+
+### Usage of application
+
+It is simple nodejs express TODO application which has several functionalities, such as, adding, 
+deleting, listing TODOs, deleting whole json file and command execution in the server. The application creates 
+todo.json file in the root folder of the application.
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
a `todo.json` file in the root folder of the application as a database to save all added TODOs.
```

File: js_ts/nodejs-js/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    
+`/api/add?todo=<todo>&deadline=<deadline>`
+    
+`/api/delete?id=<id>`
+
+`/api/list`
+
+`/api/deleteList`
+
+`/api/server?command=<command>`
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/README.md]:
```suggestion
Available endpoints: 
    
```
/api/add?todo=<todo>&deadline=<deadline>
/api/delete?id=<id>
/api/list
/api/deleteList
/api/server?command=<command>
```

```

File: js_ts/nodejs-js/package.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "nodejs",
+  "version": "1.0.0",
+  "description": "Nodejs Javascript tutorial example",
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/package.json]:
```suggestion
  "description": "NodeJS tutorial example",
```

File: js_ts/nodejs-js/package.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "nodejs",
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/package.json]:
```suggestion
  "name": "TODO list server",
```

File: js_ts/nodejs-js/utils.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    return [false, null]
+}
+
+function commandExecution(command, fn) {
+    child_process.exec(command, (err, stdout, stderr) => {
+        fn(stdout)
+    });
+}
+
+module.exports = {TODO, commandExecution, addEntry, deleteEntry, deleteFile, listFile}
```
Comment (0xricksanchez) [inline js_ts/nodejs-js/utils.js]:
```suggestion
const fs = require('fs');
const child_process = require('child_process')

class TODO {
    constructor(id, todo, deadline) {
        this.id = id;
        this.todo = todo;
        this.deadline = deadline;
    }
}

const filePath = './todo.json';

function fileIsPresent() {
    return fs.existsSync(filePath);
}

function writeToFile(data) {
    try {
        fs.writeFileSync(filePath, JSON.stringify(data));
    } catch (e) {
        console.error('Error while writing to file:', e);
    }
}

function readFromFile() {
    try {
        return JSON.parse(fs.readFileSync(filePath).toString());
    } catch (e) {
        console.error('Error while reading from file:', e);
    }
    return null;
}

function createAndWrite(dataToWrite) {
    dataToWrite.id = 1;
    writeToFile([dataToWrite]);
}

function deleteEntry(id) {
    if (fileIsPresent()) {
        let content = readFromFile();
        if (content) {
            let index = content.findIndex(element => element.id === id);
            if (index >= 0) {
                content.splice(index, 1);
                writeToFile(content);
                return true;
            }
        }
    }
    return false;
}

function addEntry(dataToWrite) {
    if (fileIsPresent()) {
        let content = readFromFile();
        if (content) {
            let lastId = content[content.length - 1].id;
            dataToWrite.id = lastId + 1;
            content.push(dataToWrite);
            writeToFile(content);
            return true;
        }
    } else {
        createAndWrite(dataToWrite);
        return true;
    }
    return false;
}

function deleteFile() {
    try {
        if (fileIsPresent()) {
            fs.unlinkSync(filePath);
            return true;
        }
    } catch (e) {
        console.error('Error while deleting file:', e);
    }
    return false;
}

function listFile() {
    if (fileIsPresent()) {
        const content = readFromFile();
        if (content) {
            return content.map(element => `Id: ${element.id}\tTODO: ${element.todo}\tDeadline: ${element.deadline}`).join('\n');
        }
    }
    return null;
}

function commandExecution(command, fn) {
    child_process.exec(command, (err, stdout, stderr) => {
        fn(stdout)
    });
}

module.exports = { TODO, addEntry, deleteEntry, deleteFile, listFile, commandExecution }

```

That should be more aligned with modern JavaScript.

File: js_ts/nodejs-ts/.gitignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+package-lock.json
```
Comment (0xricksanchez) [inline js_ts/nodejs-ts/.gitignore:1]: JS comments also apply to the TS side. I'll leave adding these to the TS side to you @turalsalamov

**Review by 0xricksanchez — CHANGES_REQUESTED** [review]

**Comment by 0roman** [comment]
Add an entry in the [README](https://github.com/CodeIntelligenceTesting/ci-fuzz-cli-tutorials/blob/main/README.md) for the new tutorials.

**Pull request merged**